Microsoft.NET.Build.Containers/AuthHandshakeMessageHandler.cs

```diff
@@ -0,0 +1,149 @@
+            {
+                authInfo = new AuthInfo(new Uri(realm), service, scope);
+                return true;
+            }
+        }
+
+        return false;
+    }
+
+    public AuthHandshakeMessageHandler(HttpMessageHandler innerHandler) : base(innerHandler) { }
+
+    /// <summary>
+    /// Response to a request to get a token using some auth.
+    /// </summary>
+    /// <remarks>
+    /// <see href="https://docs.docker.com/registry/spec/auth/token/#token-response-fields"/>
+    /// </remarks>
+    private record TokenResponse(string? token, string? access_token, int? expires_in, DateTimeOffset? issued_at)
+    {
+        public string ResolvedToken => token ?? access_token ?? throw new ArgumentException("Token response had neither token nor access_token.");
+    }
+
+    /// <summary>
+    /// Uses the authentication information from a 401 response to perform the authentication dance for a given registry.
+    /// Credentials for the request are retrieved from the credential provider, then used to acquire a token.
+    /// That token is cached for some duration on a per-host basis.
+    /// </summary>
+    /// <param name="realm"></param>
+    /// <param name="service"></param>
+    /// <param name="scope"></param>
+    /// <param name="cancellationToken"></param>
+    /// <returns></returns>
+    private async Task<string> GetTokenAsync(Uri realm, string service, string scope, CancellationToken cancellationToken)
+    {
+        // fetch creds for the host
+        DockerCredentials privateRepoCreds = await CredsProvider.GetCredentialsAsync(realm.Host);
+        // use those creds when calling the token provider
+        var header = privateRepoCreds.Username == "<token>"
+                        ? new AuthenticationHeaderValue("Bearer", privateRepoCreds.Password)
+                        : new AuthenticationHeaderValue("Basic", Convert.ToBase64String(Encoding.ASCII.GetBytes($"{privateRepoCreds.Username}:{privateRepoCreds.Password}")));
+        var builder = new UriBuilder(realm);
+        var queryDict = System.Web.HttpUtility.ParseQueryString("");
+        queryDict["service"] = service;
+        queryDict["scope"] = scope;
+        builder.Query = queryDict.ToString();
+        var message = new HttpRequestMessage(HttpMethod.Get, builder.ToString());
+        message.Headers.Authorization = header;
+
+        var tokenResponse = await base.SendAsync(message, cancellationToken);
+        tokenResponse.EnsureSuccessStatusCode();
+
+        TokenResponse? token = JsonSerializer.Deserialize<TokenResponse>(tokenResponse.Content.ReadAsStream());
+        if (token is null)
+        {
+            throw new ArgumentException("Could not deserialize token from JSON");
+        }
+
+        // save the retrieved token in the cache
+        TokenCache[realm.Host] = token.ResolvedToken;
```

Comment:
is the thought that the short duration makes the fancier cache not worth it?

Reply:
Yeah, a few thoughts here:

1. Because of https://github.com/dotnet/sdk-container-builds/pull/184/files#r984941329 it was a pain to add the dependency (it requires multiple DLLs).
2. I don't know how often the process in which this task lives will have a long-enough lifetime to care about expiration.
3. In debugging, I observed very few successful hits on cached values, because the different operations tended to require different scopes (pull vs push, different repositories/image names), so the reauth flow had to happen most of the time anyway.

Reply:
All of that makes perfect sense, thanks for confirming